src/pybind/mgr/poolsets/module.py

```diff
@@ -0,0 +1,1600 @@
+        return None
+
+    def _command_poolset_create(self, cmd):
+        # First, consider the application, and use that compose
+        # a set of intended pools.  Mark each pool as either for
+        # metadata or data, and give it a fraction of the overall
+        # space allocation (does not have to add up to 1).
+
+        class PoolIntent(object):
+            def __init__(self, application, suffix, metadata, weight):
+                # Populate these fields from user input
+                self.application = application
+                self.suffix = suffix
+                self.metadata = metadata
+                self.weight = weight
+
+                # Populate these fields from policy
+                self.initial_pg_num = None
+                self.crush_rule = None
+                self.target_size = None
+                self.target_ratio = None
+                self.name = None
+
+                # Populate these fields after creation
+                self.pool_id = None
+
+            def dump(self):
+                return {
+                    'name': self.name,
+                    'initial_pg_num': self.initial_pg_num,
+                    'crush_rule': self.crush_rule,
+                    'application': self.application
+                }
+
+            def undump(self, data):
+                self.name = data['name']
+                self.initial_pg_num = data['initial_pg_num']
+                self.crush_rule = data['crush_rule']
+                self.application = data['appliction']
+
+        pool_set_name = cmd['psname']
+        application = cmd['app']
+
+        capacity_ratio = None
+        capacity_bytes = None
+        size_str = cmd['size']
+        if size_str.endswith("%"):
+            capacity_ratio = int(size_str[0:-1]) / 100.0
+        else:
+            try:
+                capacity_bytes = parse_friendly_bytes(size_str)
+            except ValueError as e:
+                return -errno.EINVAL, "", str(e)
+
+        if pool_set_name in self._poolsets:
+            if application in self._poolsets[pool_set_name]._application:
+                # It's kind of awkward to try and fully check whether
+                # the existing poolset is the same as what we would have
+                # created, so be pretty liberal in what we accept for
+                # the idempotent case.
+                return 0, "", "Poolset '%s' already exists" % pool_set_name
+            else:
+                return -errno.EEXIST, "", "Poolset '%s' already exists" % pool_set_name
+
+        # Construct PoolIntent objects to describe the pools we'd like to create
+        if application == "rados" or application == "rbd":
+            # Single data pool
+            pool_intents = [PoolIntent(application, "", False, 1.0)]
+        elif application == "cephfs":
+            pool_intents = [PoolIntent(application, "meta", True, 0.1),
+                            PoolIntent(application, "data", False, 1.0)]
```

Comment:
Is the intent to make these pool names compatible with what the rook filesystem object expects? As it stands today, they don't seem to be compatible. If you have a filesystem named "myfs", rook currently wants myfs-metadata and myfs-data0 (and maybe -data1, -data2, etc?). This seems to produce pools named "myfs.data" and "myfs.metadata".

Reply:
Nope, not trying to mirror any existing tools with this.  Rook-created pools should be picked up in the handle_osd_map path, and if the pools are created with this module (e.g. when using the volumes module on top of it) then I'd expect Rook to just know the filesystem name, and not care about pool names (once https://github.com/rook/rook/pull/1900 is implemented)

Reply:
Ok, I think I understand the design here then, thanks.